sound/soc/sof/intel/hda.c

```diff
@@ -328,27 +605,44 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
 	sdev->hda = hdev;
 	hdev->desc = chip;
 
-	/* HDA base */
-	sdev->bar[HDA_DSP_HDA_BAR] = pci_ioremap_bar(pci, HDA_DSP_HDA_BAR);
-	if (!sdev->bar[HDA_DSP_HDA_BAR]) {
-		dev_err(&pci->dev, "error: ioremap error\n");
-		/*
-		 * FIXME: why do we return directly,
-		 *  should we have a goto err here?
-		 *  or should all these gotos be replaced
-		 * by a return?
-		 */
-		return -ENXIO;
-	}
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	hbus = devm_kzalloc(&pci->dev, sizeof(*hbus), GFP_KERNEL);
+	if (!hbus)
+		return -ENOMEM;
+
+	sdev->hbus = hbus;
+	bus = &hbus->core;
+
+	/* HDA bus init */
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDA)
+	ext_ops = snd_soc_hdac_hda_get_ops();
+#endif
+	snd_hdac_ext_bus_init(bus, &pci->dev, NULL, NULL, ext_ops);
+	bus->use_posbuf = 1;
+	bus->bdl_pos_adj = 0;
+
+	mutex_init(&hbus->prepare_mutex);
+	hbus->pci = pci;
+	hbus->mixer_assigned = -1;
+	hbus->modelname = "sofbus";
+#endif
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	/* initialise hdac bus */
+	ret = sof_hdac_first_init(sdev);
+#else
+	/* initialise resources for non-HDA */
+	ret = sof_first_init(sdev);
+#endif
+	if (ret < 0)
+		return ret;
 
 	/* DSP base */
 	sdev->bar[HDA_DSP_BAR] = pci_ioremap_bar(pci, HDA_DSP_BAR);
```

Comment:
the part below is super hard to review.

Reply:
um, Liam helps to refine and consolidate them in his consequent commit.

Reply:
Differentiation was moved to separate new functions to avoid lots of conditional compilation logic in a single function.